### PR TITLE
Prepare bounded runtime matrix validation

### PR DESCRIPTION
**Pull request by lox**
## Why

Needs-derived matrices cannot be expanded during the initial workflow compile. We need to retain and validate the narrow PostHog-shaped boundary without activating a client-side continuation protocol that cannot yet make reliable retry guarantees.

The pinned PostHog workflow builds a JSON array in [`build_django_matrix`](https://github.com/buildkite/posthog/blob/b9a0729f733e6dd170017677bb339608489ec3d2/.github/workflows/ci-backend.yml#L2251-L2264), publishes it as the `include` output, and consumes it in [`django`](https://github.com/buildkite/posthog/blob/b9a0729f733e6dd170017677bb339608489ec3d2/.github/workflows/ci-backend.yml#L2538-L2568):

```yaml
strategy:
  fail-fast: false
  matrix:
    include: ${{ fromJson(needs.build_django_matrix.outputs.include) }}
```

Previously this shape was discarded as a generic unsupported matrix expression. This PR recognizes the exact direct-output form and records enough immutable information to validate and later expand it safely.

## What

- Accept only a complete `fromJSON(needs.<direct-job>.outputs.<name>)` expression, including GitHub's case-insensitive `fromJson` spelling.
- Require one declared direct producer, one declared output, and one statically expanded producer instance; direct reusable-workflow outputs must project to one concrete flattened producer/output.
- Retain a compile-time descriptor binding the source path/digest/span, producer step key/output, consumer job, shape, and fixed limits.
- Strictly expand a matrix object or include array with bounded JSON, graph size, value types, property names, canonical number spelling, and deterministic collision checks against both generated and existing step keys.
- Preserve recognition through reusable-workflow graph failures with order-independent, bounded, fail-closed discovery, so unsupported continuation syntax cannot fall through to event-metadata acquisition.

This is preparation, not runtime activation: normal `validate`, `compile`, and `upload` remain incompatible, produce no continuation plans or pipeline bytes, and make no Buildkite upload calls. Upload rejects valid, descriptor-invalid, and early-graph-failing exact runtime-matrix boundaries before reading event metadata. Runtime upload and retry behavior remain separate work.

Related: [PB-2621](https://linear.app/buildkite/issue/PB-2621/dynamic-matrix-expansion-with-idempotent-continuation-uploads)

### CODE REVIEW

File: internal/compiler/runtime_matrix.go [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+	}
+	reference, err := expression.RuntimeMatrixOutput(*expr)
+	if err != nil {
+		return RuntimeMatrixDescriptor{}, false, nil
+	}
+	if shape == RuntimeMatrixShapeInclude && (len(job.Matrix.Rows) != 0 || job.Matrix.Include != nil || job.Matrix.Exclude != nil) {
+		return RuntimeMatrixDescriptor{}, true, errors.New("runtime matrix include output must be the complete matrix definition")
+	}
+
+	producerID, err := exactRuntimeMatrixNeed(job.Needs, reference.Job)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/compiler/runtime_matrix.go]: Fixed in c9d7dd6: runtime-matrix producer selection now resolves the source-level direct need through its flattened need binding, requires exactly one concrete producer, and records the namespaced producer identity. Added a local reusable-workflow regression test; `mise run check` passes.

File: internal/compiler/runtime_matrix.go [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+func DecodeRuntimeMatrixDescriptor(source []byte) (RuntimeMatrixDescriptor, error) {
+	if len(source) > MaxRuntimeMatrixDescriptorBytes {
+		return RuntimeMatrixDescriptor{}, fmt.Errorf("runtime matrix descriptor is %d bytes, maximum is %d", len(source), MaxRuntimeMatrixDescriptorBytes)
+	}
+	if err := rejectRuntimeMatrixDuplicateKeys(source); err != nil {
+		return RuntimeMatrixDescriptor{}, fmt.Errorf("decode runtime matrix descriptor: %w", err)
+	}
+	var descriptor RuntimeMatrixDescriptor
+	decoder := json.NewDecoder(bytes.NewReader(source))
+	decoder.DisallowUnknownFields()
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/compiler/runtime_matrix.go:266]: Fixed in c9d7dd6: descriptor decoding now rejects non-exact keys at the top level and within source spans/positions before Go’s case-insensitive struct matching can accept them. Added `Schema`, `Source`, `Start`, and `Line` rejection cases; `mise run check` passes.

File: internal/compiler/runtime_matrix.go [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+
+	producerID, err := exactRuntimeMatrixNeed(needs, reference.Job)
+	if err != nil {
+		return RuntimeMatrixDescriptor{}, true, err
+	}
+	producer, exists := jobs[producerID]
+	if !exists {
+		return RuntimeMatrixDescriptor{}, true, fmt.Errorf("runtime matrix producer %q is unavailable after graph resolution", reference.Job)
+	}
+	output, err := exactRuntimeMatrixOutput(producer.Outputs, reference.Output)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/compiler/runtime_matrix.go]: Fixed in d98c445: projected reusable-workflow outputs now resolve through the direct need’s `needBinding.outputs` entry to one exact flattened member and inner output. A regression test covers a two-job reusable call whose declared `include` output maps to `jobs.build.outputs.matrix`; `mise run check` passes.

File: internal/compiler/runtime_matrix.go [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+		return nil, fmt.Errorf("runtime continuation expands graph beyond %d jobs", MaxRuntimeMatrixGraphJobs)
+	}
+	seenValues := make(map[string]struct{}, len(matrices))
+	seenKeys := make(map[string]struct{}, len(matrices))
+	canonicalKeys := make(map[string]string)
+	for _, matrix := range matrices {
+		if err := validateExpandedRuntimeMatrix(matrix, canonicalKeys); err != nil {
+			return nil, err
+		}
+		canonical, err := json.Marshal(matrix)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/compiler/runtime_matrix.go:389]: Fixed in d98c445: accepted JSON numbers are normalized to one exact scientific spelling before duplicate detection and instance-key hashing, so `1`, `1.0`, and `1e0` all become `1`. Added canonicalization, equivalent-instance rejection, and exponent-boundary tests; `mise run check` passes.

File: internal/compiler/reusable.go [hunk 5]
```diff
@@ -198,6 +201,7 @@ func (resolver *reusableResolver) resolve(path, digest string, parsed *workflow.
 		if err != nil {
 			return reusableResolution{}, err
 		}
+		resolver.runtimeMatrixBoundary = resolver.runtimeMatrixBoundary || hasRuntimeMatrixBoundary(callee)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/compiler/reusable.go:246]: Fixed in e470642: boundary discovery now pre-scans reachable local reusable workflows independently of fail-fast graph resolution, rescans shared callees when reached at a shallower depth, and fails closed on incomplete discovery. Regressions cover failure-before-boundary ordering, shared deep/shallow reachability, and parse uncertainty with zero Buildkite calls.

File: internal/compiler/reusable.go [hunk 6]
```diff
@@ -118,7 +121,48 @@ func hasReusableCall(parsed *workflow.Workflow) bool {
 	return false
 }
 
+func (resolver *reusableResolver) discoverRuntimeMatrixBoundaries(parsed *workflow.Workflow, depth int, scannedAtDepth map[string]int) {
+	resolver.runtimeMatrixBoundary = resolver.runtimeMatrixBoundary || hasRuntimeMatrixBoundary(parsed)
+	if depth >= maxReusableWorkflowDepth {
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/compiler/reusable.go:126]: Fixed in 170f517: reaching the reusable-workflow depth limit with an unscanned call now marks discovery incomplete and rejects before event metadata. The regression builds that exact depth-limit shape and asserts zero Buildkite calls.

**Comment by lox** [comment]
Review follow-up summary:

- for review 4913235681: addressed the reported edge cases in subsequent commits and tests.
- for review 4913351148: addressed the additional projection and numeric-boundary cases in subsequent commits and tests.
- for review 4913404421: acknowledged; the reviewed revision had no further findings.
- for review 4913582818: acknowledged; the reviewed revision had no further findings.
- for review 4913862880: acknowledged; final head 170f517 fixes the last depth-limit blocker and the latest review found nothing further.

**Comment by sj26** [comment]
Released in [v0.10.1](https://github.com/buildkite/buildkite-gha/releases/tag/v0.10.1)
